### DISCUSSION TITLE
How to Add a Save Button to the Toolbar of ui.editor() in NiceGUI

### DISCUSSION THREAD

**Discussion by Daniel-Fei** [zauberzeug/nicegui#3556] [Q&A] [answered]
### Question

Hi everyone,

I'm working on a project with NiceGUI and need some help. I'm trying to add a "Save" button to the toolbar of ui.editor() so that when it's clicked, it triggers a function (e.g., saving content to a database).

I found an example on the Quasar docs (https://quasar.dev/vue-components/editor) and tried to mimic it in NiceGUI, but I haven't had any luck getting it to work. Has anyone successfully done this or have any pointers on how to achieve it?

Thanks in advance for your help!

**Comment by python-and-novella** [comment]
Do you want copy it(https://quasar.dev/vue-components/editor#example--add-new-commands)?

**Reply by Daniel-Fei** [reply]
Yes, I want to copy that example. Could you please help me on this? Thank you!

**Comment by falkoschindler** [comment]
Here's an example how it _should_ work. But somehow the `handler` doesn't seem to be called:
```py
editor = ui.editor()
editor._props['definitions'] = {
    'save': {
        'tip': 'Save your work',
        'icon': 'save',
        'label': 'Save',
        ':handler': '() => console.log("Save button clicked")',
    },
}
editor._props['toolbar'] = [['save']]
```

Ultimately the handler should emit an event which can be handled on the server. But first I wanted to trigger some JavaScript code by clicking the button.

**Reply by Daniel-Fei** [reply]
Hi @falkoschindler,
Thank you for the quick guidance! I tried the code, and the save button does appear in the toolbar. As you mentioned, how can I execute a function after clicking the button? Could you please provide further instructions? Thanks again!

**Reply by python-and-novella** [reply]
We can use slot.
Inspired by https://quasar.dev/vue-components/editor#example--using-toolbar-slots
```python3
editor = ui.editor().classes('w-[480px]')
editor._props.update(toolbar=[ [ 'left', 'center', 'right', 'justify' ], [ 'bold', 'italic', 'underline', 'strike' ], [ 'undo', 'redo' ],['save'] ])

with editor.add_slot('save'):
    ui.button(icon='save',on_click=lambda :ui.notify('file saved!')).props('flat dense').classes('text-black p-0')
```
![image](https://github.com/user-attachments/assets/85a646ad-8bc1-44ad-b399-6986ac886659)

**Reply by python-and-novella** [reply, marked as answer]
Here is the shortest code of `add save button`,but it is so ugly.
```python3
with ui.editor().props(''':toolbar="[['left','center','right','justify'],['bold','italic','underline','strike'], ['undo','redo'],['save']]"''').add_slot("save"):
    ui.button(icon="save",on_click=lambda:ui.notify("file saved!")).props("flat dense").classes("text-black p-0")
```

**Reply by falkoschindler** [reply]
Nice! It's a pity that we need to repeat all toolbar buttons to add our custom one. But the result is great! 😀
You can add "size=0.8em" to the button props to better match the size of other toolbar buttons.

**Reply by Daniel-Fei** [reply]
Thank you, @python-and-fiction and @falkoschindler. The latest code you provided works perfectly!
I appreciate the quick and very helpful assistance. Best wishes!

**Comment by Daniel-Fei** [comment]
Hi @python-and-fiction @falkoschindler,
I have another question. When I modify the code to the following:

```python
with ui.editor().props(''':toolbar="[['left','center','right','justify'],['bold','italic','underline','strike'], ['undo','redo'],['save']]"''').add_slot("save") as editor:
    ui.button(icon="save",on_click=lambda:ui.notify(editor.value)).props("flat dense size=0.8em").classes("text-black p-0")
```

I encounter an error:
'Slot' object has no attribute 'value'.

Could you please advise on how to resolve this issue?

Thank you for your assistance!

**Reply by python-and-novella** [reply]
`add_slot` returns slot object. If you want to use element object, you should use `as` before you add a slot.
It likes this:
```python3
with ui.element() as ele, ele.add_slot():
    #here belongs to slot obeject,but ele is element object you can use
```
So the correct is:
```python3
with ui.editor().props(''':toolbar="[['left','center','right','justify'],['bold','italic','underline','strike'], ['undo','redo'],['save']]"''') as editor, editor.add_slot("save"):
    ui.button(icon="save",on_click=lambda:ui.notify(editor.value)).props("flat dense size=0.8em").classes("text-black p-0")
```

**Reply by Daniel-Fei** [reply]
@python-and-fiction Thank you, that perfectly solved the issues!

I have one last question—would it be possible to add a color palette button to select text colors in a similar way? Thank you again for your guidance!

**Reply by python-and-novella** [reply]
Do you mean this:
[v1] Feature Request - QEditor - Font Color and Background Color · Issue #3269 · quasarframework/quasar
https://github.com/quasarframework/quasar/issues/3269
Or this:
```python3
with ui.editor().props(''':toolbar="[['left','center','right','justify'],['bold','italic','underline','strike'], ['undo','redo'],['viewsource','save','color1','color']]"''').classes('w-[480px]') as editor: 
    with editor.add_slot("save"):
        ui.button(icon="save",on_click=lambda:ui.notify(f'{len(editor.value)} chars saved!')).props("flat dense size=0.8em").classes("text-black p-0")
    with editor.add_slot("color1"):
        with ui.button(icon='colorize').props("flat dense size=0.8em").classes("text-black p-0"):
            ui.color_picker(on_pick=lambda e: editor.props(f'content-class="text-[{e.color}]"'))
```

**Reply by Daniel-Fei** [reply]
Hi @python-and-fiction ,

I tried your code, and it works great! It achieves the effect I was looking for.

However, I have one more issue I'd like to seek your guidance on. With the current code, when I select a color, it changes the color of all the text. Is there a way to modify it so that only the selected portion of the text changes color? For example, I would like to select a portion of the text, then click the color button and choose a color that only affects the selected text, leaving the rest of the text unchanged.

Thank you!

**Reply by python-and-novella** [reply]
I am trying to achieve it too.
It seems that `caret` of `q-editor` is the way to access seleted content.
Maybe someone could achieve it following my code.
BTW,your question is offtopic now, you'd better create a new discussion for that others can search it easily.

**Reply by Daniel-Fei** [reply]
Thank you, @python-and-fiction.

I appreciate your continued guidance. I will create a new discussion to further seek advice on this matter. Thanks again!
